File: 프로그래머스/unrated/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.py
def solution(today, terms, privacies):
    date = list(map(int,today.split(".")))
    year = date[0]
    month = date[1]
    day = date[2]
    
    dict = {}
    for term in terms:
        term_ver = term.split()[0]
        term_period = int(term.split()[1])
        dict[term_ver] = term_period
    
    answer = []
    for i in range(len(privacies)):
        privacy_date = list(map(int, privacies[i].split()[0].split(".")))
        privacy_year = privacy_date[0]
        privacy_month = privacy_date[1]
        privacy_day = privacy_date[2]
        
        if privacy_year == year:
            if privacy_day <= day:
                gap = (month - privacy_month)*28 + (day - privacy_day)
            else:
                gap = (month - privacy_month)*28- (privacy_day - day)
        else:
            if privacy_month <= month:
                if privacy_day <= day:
                    gap = (year-privacy_year)*12*28 + (month-privacy_month)*28 + (day - privacy_day)
                else:
                    gap = (year-privacy_year)*12*28 + (month-privacy_month)*28 - (privacy_day - day)
            else:
                if privacy_day <= day:
                    gap = (year-privacy_year)*12*28- (privacy_month - month)*28 + (day - privacy_day)
                else:
                    gap = (year-privacy_year)*12*28 - (privacy_month - month)*28  - (privacy_day - day)
                    
        
        
        privacy_ver = privacies[i].split()[1]
        
        if gap >= dict[privacy_ver]*28:
            answer.append(i+1)
                          
    return answer

File: 프로그래머스/unrated/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.js
function dateToNumber(date){
    let arr = date.split('.')
    let year = arr[0]
    let month = arr[1]
    let day = arr[2]
    let result = Number(day) + Number(month)*30 + (Number(year)-2000)*12*30
    return result
}

function solution(today, terms, privacies) {
    // 날짜는 2000년도를 빼준값을 사용
    // n개의 개인정보가 있음 => privacies
    // privacies 에서 today를 뺀 값이 음수면 파기해야된다. terms에 정보가 있음
    // terms 객체를 만들어줌
    let info = {}
    for (term of terms){
        let data = term.split(' ')
        let type = data[0]
        let period = data[1]
        info[type] = period
    }
    let answer = []
    let i = 1
    for (privacy of privacies){
        let data = privacy.split(' ')
        let date = data[0]
        let type = data[1]
        
        // date + type 이 today 보다 작으면 파기
        let originalDate = dateToNumber(today)
        let privacyDate = dateToNumber(date)
        let expireDate = privacyDate + Number(info[type])*30
        if (expireDate <= originalDate){
            answer.push(i)
        }
        i++
    }
    
    return answer
}